atp/client.go

```diff
@@ -356,17 +393,23 @@ func (c *client) executeReadLoop(cborReader *cbor.Decoder) {
 			var doneMessage WorkDoneMessage
 			if err := cbor.Unmarshal(runtimeMessage.RawMessageData, &doneMessage); err != nil {
 				c.logger.Errorf("Failed to decode work done message (%v) for run ID '%s' ", err, runtimeMessage.RunID)
+				c.mutex.Lock()
 				c.sendExecutionResult(runtimeMessage.RunID, NewErrorExecutionResult(
 					fmt.Errorf("failed to decode work done message (%w)", err)))
+				c.mutex.Unlock()
 			}
+			c.mutex.Lock()
 			c.sendExecutionResult(runtimeMessage.RunID, c.processWorkDone(runtimeMessage.RunID, doneMessage))
+			c.mutex.Unlock()
 		case MessageTypeSignal:
 			var signalMessage SignalMessage
 			if err := cbor.Unmarshal(runtimeMessage.RawMessageData, &signalMessage); err != nil {
 				c.logger.Errorf("ATP client for run ID '%s' failed to decode signal message: %v",
 					runtimeMessage.RunID, err)
 			}
+			c.mutex.Lock()
 			signalChannel, found := c.runningStepEmittedSignalChannels[runtimeMessage.RunID]
+			c.mutex.Unlock()
```

Comment:
This is interesting code...do we ever delete entries from `c.runningStepEmittedSignalChannels`?  That is, once we release this mutex, we might not be able to depend on the value of `signalChannel` if we do.

Reply:
They are deleted in `sendExecutionResult(...)`. Worst case scenario, if there is a race, then yes, there will be an error. Hopefully the values are transmitted in order. If that's a design flaw, that can be handled in an issue.
The main risk that I see is hanging the code by holding the mutex during that call.

Reply:
Resumed [below](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/92#discussion_r1669102859).